test/integration/tcp_tunneling_integration_test.cc

```diff
@@ -596,9 +633,175 @@ TEST_P(TcpTunnelingIntegrationTest, H2ConnectionReuse) {
+  ASSERT_TRUE(fake_upstream_connection_->waitForNewStream(*dispatcher_, upstream_request_));
+  ASSERT_TRUE(upstream_request_->waitForHeadersComplete());
+  upstream_request_->encodeHeaders(default_response_headers_, false);
+
+  ASSERT_TRUE(tcp_client2->write("hello2", false));
+  ASSERT_TRUE(upstream_request_->waitForData(*dispatcher_, "hello2"));
+  ASSERT_TRUE(fake_upstream_connection_->close());
+
+  ASSERT_TRUE(fake_upstream_connection_->waitForDisconnect());
+  tcp_client2->waitForHalfClose();
+  tcp_client2->close();
+}
+
+TEST_P(TcpTunnelingIntegrationTest, InvalidResponseHeadersHttp1) {
+  if (upstreamProtocol() == FakeHttpConnection::Type::HTTP1) {
+    return;
+  }
+  initialize();
+
+  // Start a connection, and verify the upgrade headers are received upstream.
+  IntegrationTcpClientPtr tcp_client = makeTcpConnection(lookupPort("tcp_proxy"));
+  ASSERT_TRUE(fake_upstreams_[0]->waitForHttpConnection(*dispatcher_, fake_upstream_connection_));
+  ASSERT_TRUE(fake_upstream_connection_->waitForNewStream(*dispatcher_, upstream_request_));
+  ASSERT_TRUE(upstream_request_->waitForHeadersComplete());
+
+  // Send invalid response_ headers, and verify that the client disconnects and
+  // upstream gets a stream reset.
+  default_response_headers_.setStatus(enumToInt(Http::Code::ServiceUnavailable));
+  upstream_request_->encodeHeaders(default_response_headers_, false);
+  ASSERT_TRUE(upstream_request_->waitForReset());
+
+  // The connection should be fully closed, but the client has no way of knowing
+  // that. Ensure the FIN is read and clean up state.
+  tcp_client->waitForHalfClose();
+  tcp_client->close();
+}
+
+TEST_P(TcpTunnelingIntegrationTest, ContentLengthHeaderIgnoredHttp1) {
+  if (upstreamProtocol() == FakeHttpConnection::Type::HTTP2) {
+    return;
+  }
+  initialize();
+
+  // Start a connection, and verify the upgrade headers are received upstream.
+  IntegrationTcpClientPtr tcp_client = makeTcpConnection(lookupPort("tcp_proxy"));
+  ASSERT_TRUE(fake_upstreams_[0]->waitForHttpConnection(*dispatcher_, fake_upstream_connection_));
+  ASSERT_TRUE(fake_upstream_connection_->waitForNewStream(*dispatcher_, upstream_request_));
+  ASSERT_TRUE(upstream_request_->waitForHeadersComplete());
+
+  // Send upgrade headers downstream, fully establishing the connection.
+  default_response_headers_.setStatus(enumToInt(Http::Code::IMUsed));
+  default_response_headers_.setContentLength(1024);
+  upstream_request_->encodeHeaders(default_response_headers_, false);
+
+  // Now send some data and close the TCP client.
+  ASSERT_TRUE(tcp_client->write("hello", false));
+  tcp_client->close();
+  ASSERT_TRUE(upstream_request_->waitForData(*dispatcher_, 5));
+  ASSERT_TRUE(fake_upstream_connection_->waitForDisconnect());
+}
+
+// TODO(irozzo): temporarily disabled as a protocol error is thrown when
+// transfer-encoding header is received in CONNECT responses.
+TEST_P(TcpTunnelingIntegrationTest, DISABLED_TransferEncodingHeaderIgnoredHttp1) {
```

Comment:
I disabled this at the moment because of:

https://github.com/envoyproxy/envoy/blob/d5e8634edf26dab65218e42f2acfe7c3dc95b72d/source/common/http/http1/codec_impl.cc#L757

Reply:
Yep, follow-up is fine.

Reply:
Personally, I'm not sure it's worth it. My plan was to try to address this point once this gets merged. WDYT @alyssawilk?

Reply:
I think as long as we land an immediate follow-up (you could start working on it now!) it's fine by me.

Reply:
I would tend to give more priority to [this](https://github.com/envoyproxy/envoy/pull/13293#issuecomment-701408912), but yes I will try to take care of both ;-)